core/src/main/java/jenkins/tasks/SimpleBuildStep.java

```diff
@@ -93,28 +93,76 @@ default void perform(@NonNull Run<?, ?> run, @NonNull FilePath workspace, @NonNu
 
     /**
      * Run this step.
+     * <p>
+     * This method <strong>must</strong> be overridden when this step requires a workspace context. If such a context is
+     * <em>not</em> required, it does not need to be overridden; it will then forward to
+     * {@link #perform(Run, EnvVars, TaskListener)}.
+     *
      * @param run a build this is running as a part of
      * @param workspace a workspace to use for any file operations
      * @param env environment variables applicable to this step
      * @param launcher a way to start processes
      * @param listener a place to send output
+     * @throws AbstractMethodError if this step requires a workspace context and neither this method nor {@link #perform(Run, FilePath, Launcher, TaskListener)} is overridden
      * @throws InterruptedException if the step is interrupted
      * @throws IOException if something goes wrong; use {@link AbortException} for a polite error
      * @since TODO
      */
     default void perform(@NonNull Run<?, ?> run, @NonNull FilePath workspace, @NonNull EnvVars env, @NonNull Launcher launcher,
                          @NonNull TaskListener listener) throws InterruptedException, IOException {
-        // If this is called, this must be an implementer of the previous API, in which case we call that, discarding
+        // If this does not require a workspace, defer to the version that does not take a workspace and launcher.
+        if (!this.requiresWorkspace()) {
+            this.perform(run, env, listener);
+            return;
+        }
+        // If we get here, this must be an implementer of the previous API, in which case we call that, discarding
         // the environment we were given.
         // But for that to work, that API method must have been implemented.
         if (Util.isOverridden(SimpleBuildStep.class, this.getClass(),
                 "perform", Run.class, FilePath.class, Launcher.class, TaskListener.class)) {
             this.perform(run, workspace, launcher, listener);
         } else {
-            throw new AbstractMethodError();
+            throw new AbstractMethodError("Unless a build step is marked as not requiring a workspace context, you must implement the overload of the perform() method that takes both a workspace and a launcher.");
         }
     }
 
+    /**
+     * Run this step, without a workspace context.
+     * <p>
+     * This method <strong>must</strong> be overridden when this step does not require a workspace context, and
+     * <strong>must not</strong> be called when such a context <em>is</em> required.
```

Comment:
```suggestion
     * will not be called when such a context <em>is</em> required.
```

(write from the PoV of an implementor)

Reply:
OK. I generally always write docs for the user of the method. But this change makes sense, and kind of works for the user/caller too.

Reply:
My point is that the Javadoc is mostly going to read by plugin implementors of build steps/wrappers, and this emphasis was only relevant for the two call sites (in core and `workflow-basic-steps`).